File: research-assistant/backend/app/papers/schemas/paper.py
from __future__ import annotations

from datetime import date, datetime
from typing import Any

from pydantic import (
    BaseModel,
    ConfigDict,
    Field,
    field_validator,
)


# ======================================================================
# PAPER AUTHOR RESPONSE
# ======================================================================


class PaperAuthorResponse(BaseModel):
    name: str

    author_id: int | str | None = None

    orcid: str | None = None

    email: str | None = None

    affiliation: str | None = None

    position: int | None = None


# ======================================================================
# PAPER BASE
# ======================================================================


class PaperBase(BaseModel):

    title: str

    abstract: str | None = None

    publication_date: date | None = None

    venue: str | None = None

    journal: str | None = None

    conference: str | None = None

    doi: str | None = None

    url: str | None = None

    pdf_url: str | None = None

    language: str | None = None

    keywords: list[str] = Field(
        default_factory=list
    )

    categories: list[str] = Field(
        default_factory=list
    )

    external_ids: dict[str, Any] = Field(
        default_factory=dict
    )

    citation_count: int = Field(
        default=0,
        ge=0,
    )

    reference_count: int = Field(
        default=0,
        ge=0,
    )

    source: str | None = None

    metadata: dict[str, Any] = Field(
        default_factory=dict
    )


# ======================================================================
# PAPER CREATE
# ======================================================================


class PaperCreate(PaperBase):

    authors: list[Any] = Field(
        default_factory=list
    )


# ======================================================================
# PAPER UPDATE
# ======================================================================


class PaperUpdate(BaseModel):

    title: str | None = None

    abstract: str | None = None

    authors: list[Any] | None = None

    publication_date: date | None = None

    venue: str | None = None

    journal: str | None = None

    conference: str | None = None

    doi: str | None = None

    url: str | None = None

    pdf_url: str | None = None

    language: str | None = None

    keywords: list[str] | None = None

    categories: list[str] | None = None

    external_ids: dict[str, Any] | None = None

    citation_count: int | None = Field(
        default=None,
        ge=0,
    )

    reference_count: int | None = Field(
        default=None,
        ge=0,
    )

    source: str | None = None

    metadata: dict[str, Any] | None = None


# ======================================================================
# PAPER INGEST REQUEST
# ======================================================================


class PaperIngestRequest(BaseModel):

    provider: str = Field(
        min_length=1,
        max_length=100,
    )

    external_id: str = Field(
        min_length=1,
        max_length=512,
    )

    @field_validator(
        "provider",
        "external_id",
    )
    @classmethod
    def strip_values(
        cls,
        value: str,
    ) -> str:

        value = value.strip()

        if not value:
            raise ValueError(
                "Value cannot be empty"
            )

        return value


# ======================================================================
# PAPER RESPONSE
# ======================================================================


class PaperResponse(PaperBase):

    model_config = ConfigDict(
        from_attributes=True
    )

    id: int | str

    authors: list[PaperAuthorResponse] = Field(
        default_factory=list
    )

    created_at: datetime | None = None

    updated_at: datetime | None = None

    canonical_key: str | None = None


# ======================================================================
# PAPER SUMMARY
# ======================================================================


class PaperSummary(BaseModel):

    id: int | str

    title: str

    abstract: str | None = None

    publication_date: date | None = None

    venue: str | None = None

    doi: str | None = None

    url: str | None = None

    citation_count: int = Field(
        default=0,
        ge=0,
    )


# ======================================================================
# PAPER LIST RESPONSE
# ======================================================================


class PaperListResponse(BaseModel):

    items: list[PaperResponse] = Field(
        default_factory=list
    )

    total: int = Field(
        default=0,
        ge=0,
    )

    page: int = Field(
        default=1,
        ge=1,
    )

    page_size: int = Field(
        default=20,
        ge=1,
        le=100,
    )

    pages: int = Field(
        default=0,
        ge=0,
    )

File: research-assistant/frontend/src/types/paper.ts

/* ========================================================================== */
/* Paper identifiers                                                          */
/* ========================================================================== */

export type PaperId = string;

/* ========================================================================== */
/* Authors                                                                    */
/* ========================================================================== */

export interface PaperAuthor {
  id?: number | string | null;
  name?: string | null;
  full_name?: string | null;
  author_id?: string | null;
  orcid?: string | null;
  email?: string | null;
  affiliation?: string | null;
  position?: number | null;
}

/* ========================================================================== */
/* Venue                                                                      */
/* ========================================================================== */

export interface PaperVenue {
  id?: number | string | null;
  name?: string | null;
  title?: string | null;
  display_name?: string | null;
}

/* ========================================================================== */
/* Paper                                                                      */
/* ========================================================================== */

export interface Paper {
  id: PaperId | null;

  title: string;

  abstract: string | null;

  authors: PaperAuthor[];

  publication_date: string | null;

  venue:
    | string
    | PaperVenue
    | null;

  journal: string | null;

  conference: string | null;

  doi: string | null;

  url: string | null;

  pdf_url: string | null;

  language: string | null;

  keywords: string[];

  categories: string[];

  external_ids: Record<
    string,
    string
  >;

  citation_count: number;

  reference_count: number;

  source: string | null;

  metadata: Record<
    string,
    unknown
  >;

  created_at: string | null;

  updated_at: string | null;

  canonical_key: string | null;
}

/* ========================================================================== */
/* Create                                                                     */
/* ========================================================================== */

export interface PaperCreate {
  title: string;

  abstract?: string | null;

  authors?: PaperAuthor[];

  publication_date?:
    | string
    | null;

  venue?: string | null;

  journal?: string | null;

  conference?: string | null;

  doi?: string | null;

  url?: string | null;

  pdf_url?: string | null;

  language?: string | null;

  keywords?: string[];

  categories?: string[];

  citation_count?: number;

  reference_count?: number;

  metadata?: Record<
    string,
    unknown
  >;
}

/* ========================================================================== */
/* Update                                                                     */
/* ========================================================================== */

export interface PaperUpdate {
  title?: string | null;

  abstract?: string | null;

  authors?: PaperAuthor[];

  publication_date?:
    | string
    | null;

  venue?: string | null;

  journal?: string | null;

  conference?: string | null;

  doi?: string | null;

  url?: string | null;

  pdf_url?: string | null;

  language?: string | null;

  keywords?: string[];

  categories?: string[];

  citation_count?: number;

  reference_count?: number;

  metadata?: Record<
    string,
    unknown
  >;
}

/* ========================================================================== */
/* Search parameters                                                          */
/* ========================================================================== */

/**
 * Parameters supported by:
 *
 * GET /api/v1/papers
 * GET /api/v1/papers/search
 */
export interface PaperSearchParams {
  /**
   * Free-text search.
   *
   * For /papers/search this is serialized
   * as `q`.
   *
   * For /papers it is serialized as `query`.
   */
  query?: string;

  author?: string;

  doi?: string;

  source?: string;

  /**
   * Provider name.
   *
   * Examples:
   * - arxiv
   * - openalex
   * - semantic_scholar
   * - crossref
   */
  provider?: string;

  /**
   * Provider-specific external paper ID.
   */
  provider_paper_id?: string;

  venue?: string;

  category?: string;

  year?: number;

  published_from?: string;

  published_to?: string;

  page?: number;

  page_size?: number;

  sort_by?: string;

  sort_order?:
    | "asc"
    | "desc";
}

/* ========================================================================== */
/* Search result                                                              */
/* ========================================================================== */

export interface PaperSearchResult {
  id: PaperId | null;

  title: string;

  abstract: string | null;

  authors: string[];

  publication_date: string | null;

  venue: string | null;

  doi: string | null;

  url: string | null;

  pdf_url: string | null;

  citation_count: number;

  reference_count?: number;

  source: string | null;

  relevance_score:
    | number
    | null;
}

/* ========================================================================== */
/* Search response                                                            */
/* ========================================================================== */

export interface PaperSearchResponse {
  papers: PaperSearchResult[];

  total: number;

  page: number;

  page_size: number;

  total_pages: number;
}

/* ========================================================================== */
/* List response                                                              */
/* ========================================================================== */

export interface PaperListResponse {
  items: Paper[];

  total: number;

  page: number;

  page_size: number;

  pages: number;
}

